### ISSUE TITLE
Support pyproject.toml (& poetry)

### ISSUE THREAD

**Issue by vicrep** [renovatebot/renovate#3110]
pyproject.toml: https://www.python.org/dev/peps/pep-0518/
Poetry: https://poetry.eustace.io/

https://poetry.eustace.io/docs/pyproject/

**Comment by rarkins** [comment]
Thanks @vicrep for nothing this. I'm keen to support Poetry soon too and glad to see there's at least one user already. Is there any chance you could fill out as much of this README as possible for Poetry and post it as a PR? Copy from https://raw.githubusercontent.com/renovatebot/renovate/feat/newmanager/lib/manager/newmanager/readme.md

**Comment by Hartorn** [comment]
@rarkins I think there are at least 3 users interested ;)

**Comment by rarkins** [comment]
@Hartorn great, let's do it! I think it should be much easier/faster than pipenv was, because (a) it seems a little more elegantly implemented, and (b) we already did most of the work anyway and just need to port it over

**Comment by Hartorn** [comment]
@rarkins can we still help you on that?

**Comment by rarkins** [comment]
@Hartorn absolutely, and I'd be happy to help you along the way as much as possible.

As far as I'm aware, we just need to duplicate the functionality of https://github.com/renovatebot/renovate/tree/master/lib/manager/pipenv into https://github.com/renovatebot/renovate/tree/master/lib/manager/poetry.

I suspect that the `update` function should be pretty much the same logically as pipenv's, because pipenv's update is mostly toml logic rather than pipenv.

The extract function can be as simply as necessary, e.g. parse the toml, locate the relevant sections, and extract the dependencies.

The final part is the artifacts, which again should be similar to pipenv's.

**Comment by Hartorn** [comment]
@rarkins thanks for the details, unfortunately I was not able to spare time in the last weeks, but as soon as I can, il will try to make a draft request

**Comment by sobolevn** [comment]
This is a blocker for me to migrate from dependabot to renovate.

**Comment by nchashch** [comment]
# TL;DR
It seems like `datasource/pypi` is already implemented and most likely can be reused as is.
In `extractPackageFile`, `updateDependency` functions we will have to parse TOML, which
was already implemented in Cargo manager and since `pyproject.toml` file is very similar
to `Cargo.toml` most likely cargo manager's code can be reused.

`getArtifacts` functions would have to be written from scratch, but it should be quite simple
(just a `poetry update dep1 dep2 dep3 ...` call).

Versioning uses almost the same syntax as npm (caret/tilde/wildcard/inequality requirements).
It should be possible to implement it in a way similar to cargo/composer/hashicorp managers.

Also there is a non trivial bit with "multiple constraints dependencies" i.e. specifying
separate version requirements for different python versions (see Multiple constraints dependencies section for more details).
# extractPackageFile/updateDependency
Poetry uses TOML for configuration, so it should be possible to use
existing code for extracting dependencies. For example it should be
possible to factor out TOML parsing code from `cargo` manager and
reuse it here.

`pyproject.toml` files are very similar to `Cargo.toml`
files (although they seem a bit simpler e.g. no table dependencies).
# getArtifacts
see [reference](https://poetry.eustace.io/docs/cli/#update)
There is a `poetry.lock` file. It can be updated with:
```
poetry update
```
or individual dependencies can be updated with:
```
poetry update requests toml
```
It isn't clear yet if there would be difficulties similar to Cargo package manager WRT `getArtifacts`,
like multiple versions of the same dependency in one project.
# datasource
see [reference](https://poetry.eustace.io/docs/repositories/#using-the-pypi-repository)
Poetry gets packages from PyPi and there is already a `datasource/pypi` module in renovate,
so this part seems to be already done.
# versions
see [reference](https://poetry.eustace.io/docs/versions/)
Poetry uses semver version constraints that are very similar to npm, it has:
- Caret requirements
- Tilde requirements
- Wildcard requirements
- Inequality requirements
- Exact requirements
- Multiple requirements

Most likely versioning code would have to be implemented in the same way as cargo/composer/hashicorp managers.

I could only find implementation of PEP440 versioning in the renovate source tree.
But it should be pretty easy to implement poetry versioning based on `versioning/semver` module,
it seem like it would be easier than cargo-semver because carets and bare versions behave like in npm.

## Multiple constraints dependencies
see [reference](https://poetry.eustace.io/docs/versions/#multiple-constraints-dependencies)
It is possible to specify a python version a dependency is compatible with like:
```toml
[tool.poetry.dependencies]
pathlib2 = { version = "^2.2", python = "~2.7" }
```
this feature doesn't seem relevant but, it is also possible to specify
separate version requirements for different python versions like:
```toml
[tool.poetry.dependencies]
foo = [
    {version = "<=1.9", python = "^2.7"},
    {version = "^2.0", python = "^3.4"}
]
```
these kinds of dependency requirements either would have to be ignored
or handled in some reasonable way.

Also note that `python=<version>` is a version requirement supporting carets/tildes/wildcards/... .
# Configuration files
see [reference](https://poetry.eustace.io/docs/pyproject/)
A poetry project is configured with `pyproject.toml` file, and
there is a `poetry.lock` file.

**Comment by rarkins** [comment]
Can anyone suggest some public repositories using poetry to fork and test against?

**Comment by ChristianMurphy** [comment]
> Can anyone suggest some public repositories using poetry to fork and test against?

https://github.com/ambv/black
https://github.com/sdispater/poetry
https://github.com/nornir-automation/nornir
